spec/ics-3-connection-semantics/README.md

```diff
@@ -0,0 +1,346 @@
+
+IBC subprotocols are reasoned about as interactions between two chains `A` and `B` — there is no prior distinction between these two chains and they are assumed to be executing the same, correct IBC protocol. `A` is simply by convention the chain which goes first in the subprotocol and `B` the chain which goes second. Protocol definitions should generally avoid including `A` and `B` in variable names to avoid confusion (as the chains themselves do not know whether they are `A` or `B` in the protocol).
+
+This ICS defines four subprotocols: opening handshake, header tracking, closing handshake, and closing by equivocation.
+
+##### Opening Handshake
+
+The opening handshake subprotocol serves to initialize roots of trust for two chains on each other and negotiate an agreeable connection version.
+
+This subprotocol need not be permissioned, modulo anti-spam measures.
+
+```golang
+type ConnOpenInit struct {
+  // Identifier to use for connection on chain A
+  Identifier  identifier
+  // Desired identifier to use for connection on chain B
+  Identifier  desiredCounterpartyIdentifier
+  // Desired version for connection
+  Version     desiredVersion
+  // Root-of-trust for chain B
+  RootOfTrust rootOfTrust
+}
+```
+
+```golang
+type ConnOpenTry struct {
+  // Desired identifier to use for connection on chain B
+  Identifier        desiredIdentifier
+  // Identifier for connection on chain A
+  Identifier        counterpartyIdentifier
+  // Desired version for connection
+  Version           desiredVersion
+  // Root-of-trust for chain A
+  RootOfTrust       rootOfTrust
+  // Proof of stored INIT state on chain A
+  AccumulatorProof  proofInit
+}
+```
+
+```golang
+type ConnOpenAck struct {
+  // Identifier for connection on chain A
+  Identifier        identifier
+  // Identifier for connection on chain B
+  Identifier        agreedCounterpartyIdentifier
+  // Agreed version for connection
+  Version           agreedVersion
+  // Proof of stored TRY state on chain B
+  AccumulatorProof  proofTry
+}
+```
+
+```golang
+type ConnOpenConfirm struct {
+  // Identifier for connection on chain B
+  Identifier        identifier
+  // Proof of stored OPEN state on chain A
+  AccumulatorProof  proofAck
+}
+```
+
+```coffeescript
+function handleConnOpenInit(identifier, desiredVersion, desiredCounterpartyIdentifier, rootOfTrust)
+  assert(Get(identifier) == null)
+  state <- INIT
+  Set(identifier, (state, desiredVersion, desiredCounterpartyIdentifier, rootOfTrust))
+```
+
+```coffeescript
+function handleConnOpenTry()
+  expectedRootOfTrust <- getRootOfTrust()
+  assert(verify(rootOfTrust, proofInit,
+    (counterpartyIdentifier, (INIT, desiredVersion, desiredIdentifier, expectedRootOfTrust))))
+  identifier <- chooseIdentifier(desiredIdentifier, counterpartyIdentifier)
```

Comment:
What does the chooseIdentifier() do? Selects one of the two? Is this the same "identifier" that is Set further down? 
To be honest I am a bit lost at this point as these two IDs seem to be for two different chains. Maybe an example could help. Also I am not sure how ID collisions are prevented.

Reply:
I have simplified the identifier usage here, now we just assume they are meaningless and prevent collisions.